### PR TITLE
[#11630] [#11631] Update conditions for session closing and closed emails

### LINKED ISSUE #11630: Session closing soon email should be sent to all students/instructors regardless of submission status
#10901 introduces skip for "closing soon" email for those who have already submitted responses.

This is a very problematic and flawed "optimization" because it took away a critical feature. It checks users for "submitting responses", not "completed all questions". Even users who have previously completed all questions may want to change their existing responses before the due time!

This so-called "optimization" needs to be undone.

### LINKED ISSUE #11631: Session closed email should only be sent to course co-owners
https://github.com/TEAMMATES/teammates/blob/9cc2e8451096b900b405d8c698291adf0e782e09/src/main/java/teammates/logic/api/EmailGenerator.java#L454-L496

Current: session closed emails are sent only to students/instructors who did not submit any response.

Problem: such email is redundant, because knowing that "the session has closed [and I have not submitted anything]" does not translate to any actionable item for the email recipients.

Solution: send session closed email only to course co-owners, who likely have follow-up actions such as publishing the result of the session.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #11630, reverts #10906
Fixes #11631, technically reverts #10914

Resolving two issues at once as they are very tightly related.

It is interesting how we have dedicated methods only to check if session has been attempted, only for them to be extremely counter-productive.

**Outline of Solution**

- Closing emails are always sent regardless of submission status. The function is now a close copy of the opening email rather than closed email, so the common function is now `generateFeedbackSessionOpeningOrClosingEmails`.
- Closed emails are repurposed to be sent to co-owners only. It is now a close copy of the opening soon email, thus another common function `generateFeedbackSessionOpeningSoonOrClosedEmails` is created.
  - The "session closed" email right now looks very empty, though. We may want to add some additional notes, although I'm not sure what.
    Screenshot 2022-03-07 at 2 31 58 AM
  - For comparison purpose, this is how "session opening soon" email looks like:
    Screenshot 2022-03-07 at 2 31 00 AM

### CODE REVIEW

File: src/main/java/teammates/logic/api/EmailGenerator.java [hunk 1]
```diff
@@ -90,72 +90,75 @@ public static EmailGenerator inst() {
                 ? instructorsLogic.getInstructorsForCourse(session.getCourseId())
                 : new ArrayList<>();
```
Comment (moziliar) [inline src/main/java/teammates/logic/api/EmailGenerator.java:91]: Can we get coowner from instructors in `generateFeedbackSessionEmailBases`?

**Review by moziliar — APPROVED** [review]
LGTM with one comment unrelated to the change.

File: src/main/java/teammates/logic/api/EmailGenerator.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/logic/api/EmailGenerator.java:91]: I would say it's possible

**Comment by wkurniawan07** [comment]
@damithc any suggestion on what the new session closed email should contain?

**Comment by damithc** [comment]
> @damithc any suggestion on what the new session closed email should contain?

Something like the below? Feel free to tweak.

Hi X,

You are receiving this email because you are a co-owner of the TEAMMATES course **[CS101] Intro to Computing**.
We wanted to let you know that the session **Peer evaluations 1** in the above course was closed at 2.30pm today, after the expiry of its deadline + grace period.
Respondents will still be able to view the submission page (and the responses they submitted earlier, if any) but they will not be able to edit existing responses or submit new responses.

Happy teaching!

TEAMMATES Team

**Comment by wkurniawan07** [comment]
@damithc I'm not saying they need to be similar, but if you compare the suggestion with the opening soon email, the opening soon email now sounds like a hurried feature (is it?).

And if we don't have the box containing the session information, it really sticks out like a sore thumb, considering all other session emails have it.

**Comment by damithc** [comment]
> @damithc I'm not saying they need to be similar, but if you compare the suggestion with the opening soon email, the opening soon email now sounds like a hurried feature (is it?).
> 
> And if we don't have the box containing the session information, it really sticks out like a sore thumb, considering all other session emails have it.

Good points. How about the below? Perhaps also add a link to go directly to the results page?
We can also tweak the 'opening' email to improve consistency, if necessary.


Hello X,

We wanted to let you know that the following session is now closed.

{info box goes here}

Notes:
* You are receiving this email because you are a co-owner of the above course.
* Respondents will still be able to view the submission page (and the responses they submitted earlier, if any) but they will not be able to edit existing responses or submit new responses.
* To view results of the session, visit ...

TEAMMATES Team

**Comment by wkurniawan07** [comment]
It looks a bit less empty now.

Screenshot 2022-03-08 at 10 44 53 PM

> We can also tweak the 'opening' email to improve consistency, if necessary.

I don't wish to introduce any unnecessary unrelated functional change.

**Pull request merged**